### DISCUSSION TITLE
Restricting volumetric kernel to elements on boundary

### DISCUSSION THREAD

**Discussion by NateAM** [idaholab/moose#28051] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

I am trying to simulate the change in volume due to the deposition or removal of mass on the surface of a deforming mesh. In my formulation it is easier for me to describe the change in volume due to a volumetric mass source term i.e.

$\frac{\partial \rho}{\partial t} + \left(\rho v_i\right)_{,i} = c$

where $c$ is the volumetric mass source term. Because for my problem the mass change through the boundary is defined on the surface as an `IntegratedBC` and I don't have any way to make this conversion directly, I would like to introduce a variable that represents a volumetric mass change for the elements on the boundary that results in the same mass change as the surface integral i.e.

$\int_{\mathcal{B}} c dv = \int_{\partial \mathcal{B}} c^s da $

where $c^s$ is the mass flow into the body across the surface. I think the answer I would expect would be for the $c$ term to be zero at all of the nodes except for on the surface that is seeing mass deposition or removal.

Is what I want to do sensible or is there a better approach? Is there a way to restrict a volumetric kernel to only contribute to the residuals on the surface of the body?

**Comment by grmnptr** [comment]
A naive approach could be to create a subdomain for the layer at the boundary and restrict your custom kernel to those elements.
A kernel in MOOSE is a block-restrictible object so you will always be able to restrict its action to a given subdomain. Usually it goes like this:
```
[Kernels]
  [mykernel]
    type = MyCustomKernel
    ... other parameters ...
    block = 'sudomainname-or-number'
  []
[]
```

**Reply by NateAM** [reply]
@grmnptr thank you for the suggestion! I would prefer not having to define new blocks if I can avoid it unless there are ways to automatically move elements from one block to another based on sideset updates. I was thinking I could try and define a monomial variable that would only be non-zero on the surface so I'll try that today and see if I can't make something work.

**Reply by GiudGiud** [reply]
> sideset updates

before the simulation or during the simulation?

**Reply by NateAM** [reply]
> before the simulation or during the simulation?

Ideally during the simulation. There will be elements that could be deactivating when there is enough mass loss so my plan is to use the `CoupledVarThresholdElementSubdomainModifier` userobject ([link to documentation](https://mooseframework.inl.gov/source/userobjects/CoupledVarThresholdElementSubdomainModifier.html))

**Reply by GiudGiud** [reply]
ok the `CoupledVarThresholdElementSubdomainModifier` is able to track the movement of the sideset as elements are added or removed.

I think you should try to:
- have a variable tracking the mass everywhere
- in the volume it solves d rho/dt + rho v = 0
- on the sideset there is a boundary condition that adds or removes mass, like an Integrated Neumann BC
- the magnitude of that flux is set proportional to the element_volume / side_area. This will mimick a volumetric term

**Reply by NateAM** [reply]
@GiudGiud okay, I think that is similar to what I have implemented though I was applying the mass change through an integrated bc on the balance of mass. If I'm missing your point please let me know!

The issue I'm running into is that I get nonphysical oscillations in the density field. To try and show things visually, in 1D where the integration points are the xs and the element nodes are the |s and I've applied a mass flux (mass per unit area) out of the body to the right hand side (->).
      e1         e2         e3
|--x--x--|--x--x--|--x--x--|->

The oscillation directions are +, -, +, - under the nodes going from left to right.

The oscillations arise because for a 1D balance of mass, this has a tri-diagonal Jacobian 

Because the surface is eroding I want to make sure that the volume-change is localized in the element nearest to the mass flux (e3 in the example). The mass of the nodes associated with e3 will be decreasing, but if the volume change is consistent with the change in mass then the density will remain constant. To do this, I need to know what the mass change rates are at the integration points of e3 which feed into the volume change kernels (balance of linear momentum stress divergence with a stress-free deformation associated with mass change).

Does that make sense?

For the mass variable how were you thinking this would be defined?

Thank you for your help!

EDIT: Formatting

**Reply by GiudGiud** [reply]
I think finding a solution in 1D is a good idea. 

> For the mass variable how were you thinking this would be defined?

You can try an elemental (constant monomial with DG or finite volume for example) or a nodal variable (default lagrange, CG). I dont know which one will be easiest I have never done this before

>  The mass of the nodes associated with e3 will be decreasing, but if the volume change is consistent with the change in mass then the density will remain constant.

this is also a good test case. Show that you can vary the size of the element (using an imposed disp_x) and still have stability of the density / mass
